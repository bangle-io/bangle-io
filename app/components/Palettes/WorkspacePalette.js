import React, { useCallback, useEffect, useMemo } from 'react';
import { Palette } from '../../ui/Palette';

import { SideBarRow } from '../Aside/SideBarRow';
import { useWorkspaces } from 'bangle-io/app/workspace2/workspace-hooks';
import { useHistory } from 'react-router-dom';

const LOG = false;

let log = LOG ? console.log.bind(console, 'play/file-palette') : () => {};

export function WorkspacePalette({ counter, query, execute, onDismiss }) {
  const { workspaces } = useWorkspaces();
  const items = useMemo(
    () =>
      workspaces.filter((ws) => {
        return strMatch(ws, query);
      }),
    [workspaces, query],
  );
  const history = useHistory();

  const onExecuteItem = useCallback(
    async (activeItemIndex) => {
      activeItemIndex =
        activeItemIndex == null
          ? Palette.getActiveIndex(counter, items.length)
          : activeItemIndex;

      const workspace = items[activeItemIndex];

      if (!workspace) {
        return;
      }

      history.push('/ws/' + workspace);
      onDismiss();
    },
    [counter, items, onDismiss, history],
  );

  useEffect(() => {
    // parent signals execution by setting execute to true
    // and expects the child to call dismiss once executed
    if (execute === true) {
      onExecuteItem();
    }
  }, [execute, onExecuteItem]);

  return items.map((item, i) => (
    <SideBarRow
      key={i}
      isActive={Palette.getActiveIndex(counter, items.length) === i}
      title={item}
      onClick={() => onExecuteItem(i)}
    />
  ));
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase().trim();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase().trim();
  return a.includes(b) || b.includes(a);
}
