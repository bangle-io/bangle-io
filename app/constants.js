export const SPLIT_SCREEN_MIN_WIDTH = parseInt(
  getComputedStyle(document.documentElement).getPropertyValue(
    '--wide-screen-min-width',
  ),
  10,
);
