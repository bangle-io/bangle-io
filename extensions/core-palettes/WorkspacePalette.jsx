import React, { useCallback, useImperativeHandle, useMemo } from 'react';
import { useWorkspaces } from 'workspaces';
import {
  AlbumIcon,
  CloseIcon,
  PaletteInfo,
  PaletteInfoItem,
} from 'ui-components/index';
import { extensionName } from './config';
import {
  MagicPaletteItem,
  MagicPaletteItemsContainer,
} from 'magic-palette/index';
import { keybindings, keyDisplayValue } from 'config/index';
import { useRecencyWatcher } from './hooks';

const LOG = false;

let log = LOG ? console.log.bind(console, 'play/file-palette') : () => {};

const identifierPrefix = 'ws:';
export const workspacePalette = {
  type: extensionName + '/workspace',
  icon: (
    <span className="pr-2 flex items-center">
      <AlbumIcon className="h-5 w-5" />
    </span>
  ),
  identifierPrefix,
  placeholder: 'Select a workspace to open',
  keybinding: keybindings.toggleWorkspacePalette.key,
  parseRawQuery: (rawQuery) => {
    if (identifierPrefix && rawQuery.startsWith(identifierPrefix)) {
      return rawQuery.slice(3);
    }
    return null;
  },
  ReactComponent: React.forwardRef(WorkspacePaletteUIComponent),
};
const storageKey = 'WorkspacePaletteUIComponent/1';
function WorkspacePaletteUIComponent(
  { query, dismissPalette, paletteItemProps },
  ref,
) {
  const { workspaces, switchWorkspace, deleteWorkspace } = useWorkspaces();
  const { injectRecency, updateRecency } = useRecencyWatcher(storageKey);

  const items = useMemo(() => {
    const _items = injectRecency(
      workspaces
        .filter((ws) => {
          return strMatch(ws.name, query);
        })
        .map((workspace, i) => {
          return {
            uid: `${workspace.name}-(${workspace.type})`,
            title: workspace.name,
            extraInfo: workspace.type,
            data: { workspace },
          };
        }),
    );

    return _items;
  }, [query, workspaces, injectRecency]);

  const onExecuteItem = useCallback(
    (getUid, sourceInfo) => {
      const uid = getUid(items);
      const item = items.find((item) => item.uid === uid);
      if (item) {
        switchWorkspace(item.data.workspace.name, sourceInfo.metaKey);
        updateRecency(uid);
      }
    },
    [switchWorkspace, updateRecency, items],
  );

  useImperativeHandle(
    ref,
    () => ({
      onExecuteItem,
    }),
    [onExecuteItem],
  );

  return (
    <>
      <MagicPaletteItemsContainer>
        {items.map((item) => {
          return (
            <MagicPaletteItem
              key={item.uid}
              items={items}
              title={item.title}
              extraInfo={item.extraInfo}
              showDividerAbove={item.showDividerAbove}
              uid={item.uid}
              isDisabled={item.disabled}
              rightIcons={item.rightIcons}
              {...paletteItemProps}
              rightHoverIcons={
                <CloseIcon
                  style={{
                    height: 16,
                    width: 16,
                  }}
                  onClick={async (e) => {
                    e.stopPropagation();
                    if (
                      window.confirm(
                        `Are you sure you want to remove "${item.data.workspace.name}"? Removing a workspace does not delete any files inside it.`,
                      )
                    ) {
                      await deleteWorkspace(item.data.workspace.name);
                      dismissPalette();
                    }
                  }}
                />
              }
            />
          );
        })}
      </MagicPaletteItemsContainer>
      <PaletteInfo>
        <PaletteInfoItem>use:</PaletteInfoItem>
        <PaletteInfoItem>
          <kbd className="font-normal">↑↓</kbd> Navigate
        </PaletteInfoItem>
        <PaletteInfoItem>
          <kbd className="font-normal">Enter</kbd> Open a workspace
        </PaletteInfoItem>
        <PaletteInfoItem>
          <kbd className="font-normal">{keyDisplayValue('Mod')}-Enter</kbd> Open
          a in new tab
        </PaletteInfoItem>
      </PaletteInfo>
    </>
  );
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase().trim();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase().trim();
  return a.includes(b) || b.includes(a);
}
