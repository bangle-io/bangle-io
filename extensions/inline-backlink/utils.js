export function removeMdExtension(str) {
  if (str.endsWith('.md')) {
    return str.slice(0, -3);
  }
  return str;
}
