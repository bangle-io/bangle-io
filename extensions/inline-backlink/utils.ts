import {
  filePathToWsPath,
  removeExtension,
  resolvePath,
  sanitizeFilePath,
  suffixWithNoteExtension,
  validateWsPath,
} from '@bangle.io/ws-path';

export function wsPathFromQuery(query: string, wsName: string) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = suffixWithNoteExtension(filePath);
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}

export function getBacklinkPath(wsPath: string, allWsPaths: readonly string[]) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeExtension(filePath);
  }

  return removeExtension(fileName);
}
