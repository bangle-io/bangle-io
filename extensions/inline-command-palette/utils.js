export function replaceSuggestText(newText) {
  return (state, dispatch, view) => {};
}
