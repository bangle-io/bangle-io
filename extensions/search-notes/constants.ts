import type { SearchResultItem } from '@bangle.io/search-pm-node';

export const CONCURRENCY = 10;

export const SHOW_SEARCH_SIDEBAR_OPERATION =
  'operation::@bangle.io/search-notes:show-search-sidebar';

export const SEARCH_SIDEBAR_NAME =
  'sidebar::@bangle.io/search-notes:search-notes';
export const extensionName = '@bangle.io/search-notes';
export interface SearchNotesExtensionState {
  searchQuery: string;
  pendingSearch: boolean;
  searchResults: SearchResultItem[] | null;
  // a counter keeping track of external change to the search query
  // useful for resetting any local state
  externalInputChange: number;
}

export type SearchNotesActions =
  | {
      name: `action::@bangle.io/search-notes:update-state`;
      value: Partial<SearchNotesExtensionState>;
    }
  | {
      name: 'action::@bangle.io/search-notes:input-search-query';
      value: {
        query: string;
      };
    }
  | {
      name: 'action::@bangle.io/search-notes:external-search-query-update';
      value: {
        query: string;
      };
    };
