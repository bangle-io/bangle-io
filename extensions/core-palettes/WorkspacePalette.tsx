import React, {
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useState,
} from 'react';

import {
  internalApi,
  nsmApi2,
  useSerialOperationContext,
} from '@bangle.io/api';
import {
  CORE_OPERATIONS_REMOVE_ACTIVE_WORKSPACE,
  CorePalette,
} from '@bangle.io/constants';
import type { WorkspaceInfo } from '@bangle.io/shared-types';
import type { PaletteOnExecuteItem } from '@bangle.io/ui-components';
import {
  AlbumIcon,
  CloseIcon,
  UniversalPalette,
} from '@bangle.io/ui-components';
import { keyDisplayValue } from '@bangle.io/utils';
import { createWsName } from '@bangle.io/ws-path';

import type { ExtensionPaletteType } from './config';
import { useRecencyWatcher } from './hooks';

const LOG = false;

let log = LOG ? console.log.bind(console, 'play/file-palette') : () => {};

const identifierPrefix = 'ws:';

const storageKey = 'WorkspacePaletteUIComponent/1';

const WorkspacePaletteUIComponent: ExtensionPaletteType['ReactComponent'] =
  React.forwardRef(
    ({ query, dismissPalette, onSelect, getActivePaletteItem }, ref) => {
      const { injectRecency, updateRecency } = useRecencyWatcher(storageKey);

      const [workspaces, updateWorkspaces] = useState<WorkspaceInfo[]>([]);

      const { dispatchSerialOperation } = useSerialOperationContext();
      useEffect(() => {
        internalApi.workspace.readAllWorkspacesInfo().then((wsInfos) => {
          updateWorkspaces(wsInfos);
        });
      }, []);

      const items = useMemo(() => {
        const _items = injectRecency(
          workspaces
            .filter((ws) => {
              return strMatch(ws.name, query);
            })
            .map((workspaceObj, i) => {
              return {
                uid: workspaceObj.name,
                title: workspaceObj.name,
                extraInfo: workspaceObj.type,
                data: { workspace: workspaceObj },
                rightHoverNode: (
                  <CloseIcon
                    style={{
                      height: 16,
                      width: 16,
                    }}
                    onClick={async (e: React.MouseEvent<any>) => {
                      e.stopPropagation();
                      dispatchSerialOperation({
                        name: CORE_OPERATIONS_REMOVE_ACTIVE_WORKSPACE,
                        value: workspaceObj.name,
                      });
                      dismissPalette();
                    }}
                  />
                ),
              };
            }),
        );

        return _items;
      }, [
        query,
        dismissPalette,
        dispatchSerialOperation,
        workspaces,
        injectRecency,
      ]);

      const activeItem = getActivePaletteItem(items);

      const onExecuteItem = useCallback<PaletteOnExecuteItem>(
        (getUid, sourceInfo) => {
          const uid = getUid(items);
          const item = items.find((item) => item.uid === uid);

          if (item && uid != null) {
            const wsName = createWsName(item.data.workspace.name);
            nsmApi2.workspace.goToWorkspace({
              wsName,
              type: sourceInfo.metaKey ? 'newTab' : 'replace',
            });

            updateRecency(uid);
          }
        },
        [updateRecency, items],
      );

      useImperativeHandle(
        ref,
        () => ({
          onExecuteItem,
        }),
        [onExecuteItem],
      );

      return (
        <>
          <UniversalPalette.PaletteItemsContainer>
            {items.map((item) => {
              return (
                <UniversalPalette.PaletteItemUI
                  key={item.uid}
                  item={item}
                  onClick={onSelect}
                  isActive={item === activeItem}
                />
              );
            })}
          </UniversalPalette.PaletteItemsContainer>
          <UniversalPalette.PaletteInfo>
            <UniversalPalette.PaletteInfoItem>
              use:
            </UniversalPalette.PaletteInfoItem>
            <UniversalPalette.PaletteInfoItem>
              <kbd className="font-normal">↑↓</kbd> Navigate
            </UniversalPalette.PaletteInfoItem>
            <UniversalPalette.PaletteInfoItem>
              <kbd className="font-normal">Enter</kbd> Open a workspace
            </UniversalPalette.PaletteInfoItem>
            <UniversalPalette.PaletteInfoItem>
              <kbd className="font-normal">{keyDisplayValue('Mod')}-Enter</kbd>{' '}
              Open a in new tab
            </UniversalPalette.PaletteInfoItem>
          </UniversalPalette.PaletteInfo>
        </>
      );
    },
  );

function strMatch(a: string[] | string, b: string): boolean {
  b = b.toLocaleLowerCase().trim();

  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase().trim();

  return a.includes(b) || b.includes(a);
}

export const workspacePalette: ExtensionPaletteType = {
  type: CorePalette.Workspace,
  icon: <AlbumIcon />,
  identifierPrefix,
  placeholder: 'Select a workspace to open',
  parseRawQuery: (rawQuery) => {
    if (identifierPrefix && rawQuery.startsWith(identifierPrefix)) {
      return rawQuery.slice(3);
    }

    return null;
  },
  ReactComponent: WorkspacePaletteUIComponent,
};
