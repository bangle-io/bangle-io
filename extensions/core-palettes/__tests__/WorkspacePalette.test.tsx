/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { act, render } from '@testing-library/react';
import React from 'react';

import { WorkspaceType } from '@bangle.io/constants';
import type { WorkspaceInfo } from '@bangle.io/shared-types';
import { sleep } from '@bangle.io/utils';

import { workspacePalette } from '../WorkspacePalette';
import { setupTestExtension } from '@bangle.io/test-utils-2';
import corePalettes from '../index';

let abortController = new AbortController();

beforeEach(() => {
  abortController = new AbortController();
});

afterEach(async () => {
  abortController.abort();
});

async function setup() {
  const ctx = setupTestExtension({
    extensions: [corePalettes],
    abortSignal: abortController.signal,
  });

  return ctx;
}

test('Component renders correctly when multiple workspaces', async () => {
  const ctx = await setup();

  const wsName = 'test-ws1';

  await ctx.createWorkspace(wsName);

  await ctx.createWorkspace(`test-ws2`);

  const dismissPalette = jest.fn();
  const onSelect = jest.fn();
  const getActivePaletteItem = jest.fn();
  const { container } = render(
    <ctx.ContextProvider>
      <workspacePalette.ReactComponent
        query=""
        paletteType={undefined}
        paletteMetadata={{}}
        updatePalette={() => {}}
        counter={0}
        updateCounter={() => {}}
        dismissPalette={dismissPalette}
        onSelect={onSelect}
        getActivePaletteItem={getActivePaletteItem}
        allPalettes={[]}
      />
    </ctx.ContextProvider>,
  );

  const prom = sleep(50);

  await act(() => prom);

  expect(container.innerHTML.includes('test-ws1')).toBe(true);
  expect(container.innerHTML.includes('test-ws2')).toBe(true);

  expect(
    container.querySelector('.B-ui-components_universal-palette-item'),
  ).toMatchSnapshot();
});

test('Component renders correctly when no workspaces', async () => {
  const ctx = await setup();
  await sleep(10);

  const dismissPalette = jest.fn();
  const onSelect = jest.fn();
  const getActivePaletteItem = jest.fn();
  const { container } = render(
    <ctx.ContextProvider>
      <workspacePalette.ReactComponent
        query=""
        paletteType={undefined}
        paletteMetadata={{}}
        updatePalette={() => {}}
        counter={0}
        updateCounter={() => {}}
        dismissPalette={dismissPalette}
        onSelect={onSelect}
        getActivePaletteItem={getActivePaletteItem}
        allPalettes={[]}
      />
    </ctx.ContextProvider>,
  );

  const prom = sleep(50);

  await act(() => prom);

  expect(container.innerHTML.includes('helpfs')).toBe(true);

  expect(
    container.querySelector('.B-ui-components_universal-palette-item'),
  ).toMatchSnapshot();
});
