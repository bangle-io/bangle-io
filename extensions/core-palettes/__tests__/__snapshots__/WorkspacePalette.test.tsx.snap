// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly 1`] = `
<div>
  <div
    class=""
    style="overflow-y: scroll;"
  >
    <div
      class="B-ui-components_universal-palette-item BU_allow-hover "
      data-id="test-ws1-(nativefs)"
      style="cursor: pointer; display: flex; justify-content: space-between; user-select: none;"
    >
      <div
        class="flex flex-row"
      >
        <div
          class="B-ui-components_left-node"
        />
        <span>
          <span
            class="text-base font-normal"
          >
            test-ws1
          </span>
          <span
            class="B-ui-components_extra-info text-base font-light"
          >
            nativefs
          </span>
        </span>
      </div>
      <div
        class="flex flex-row"
      >
        <span
          class="B-ui-components_right-node"
        />
        <span
          class="B-ui-components_right-hover-node"
        >
          <svg
            class="fill-current  "
            style="height: 16px; width: 16px;"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <div
    class="flex flex-row justify-center my-1 space "
  >
    <span
      class="text-xs ml-2 mr-2 font-light "
    >
      use:
    </span>
    <span
      class="text-xs ml-2 mr-2 font-light "
    >
      <kbd
        class="font-normal"
      >
        ↑↓
      </kbd>
       Navigate
    </span>
    <span
      class="text-xs ml-2 mr-2 font-light "
    >
      <kbd
        class="font-normal"
      >
        Enter
      </kbd>
       Open a workspace
    </span>
    <span
      class="text-xs ml-2 mr-2 font-light "
    >
      <kbd
        class="font-normal"
      >
        Ctrl
        -Enter
      </kbd>
       
      Open a in new tab
    </span>
  </div>
</div>
`;
